witchcraft/internal/tcpjson/tcp_logger.go

```diff
@@ -0,0 +1,163 @@
+
+package tcpjson
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"net"
+	"sync"
+
+	werror "github.com/palantir/witchcraft-go-error"
+	"github.com/rs/zerolog"
+)
+
+var _ io.Writer = (*TCPWriter)(nil)
+
+const errWriterClosed = "writer is closed"
+
+// envelopeSerializerFunc provides a way to change the serialization method for the provided payload.
+type envelopeSerializerFunc func(payload []byte) ([]byte, error)
+
+// TCPWriter writes logs to a TCP socket and wraps them with envelope metadata.
+type TCPWriter struct {
+	provider           ConnProvider
+	envelopeSerializer envelopeSerializerFunc
+
+	// closedChan is used to signal that the writer is shutting down
+	closedChan chan struct{}
+
+	mu   sync.RWMutex // guards conn below
+	conn net.Conn
+}
+
+func NewTCPWriter(metadata LogEnvelopeMetadata, provider ConnProvider) *TCPWriter {
+	return newTCPWriterInternal(provider, zerologSerializer(metadata))
+}
+
+func newTCPWriterInternal(provider ConnProvider, serializerFunc envelopeSerializerFunc) *TCPWriter {
+	return &TCPWriter{
+		envelopeSerializer: serializerFunc,
+		provider:           provider,
+		closedChan:         make(chan struct{}),
+		conn:               nil,
+	}
+}
+
+func (d *TCPWriter) Write(p []byte) (int, error) {
+	if d.closed() {
+		return 0, werror.Error(errWriterClosed)
+	}
+
+	envelope, err := d.envelopeSerializer(p)
+	if err != nil {
+		return 0, werror.Wrap(err, "failed to serialize the envelope")
+	}
+
+	conn, err := d.getConn()
+	if err != nil {
+		return 0, err
+	}
+
+	var total int
+	for total < len(envelope) {
+		n, err := conn.Write(envelope[total:])
+		total += n
+		if err != nil {
+			if nerr, ok := err.(net.Error); !(ok && (nerr.Temporary() || nerr.Timeout())) {
+				// permanent error so close the connection
+				_ = d.closeConn()
+				return total, err
+			}
+			return total, err
+		}
+	}
+	return len(p), nil
+}
+
+func (d *TCPWriter) getConn() (net.Conn, error) {
+	// Fast path: connection is already established and cached
+	d.mu.RLock()
+	conn := d.conn
+	d.mu.RUnlock()
+	if d.conn != nil {
+		return conn, nil
+	}
+
+	// No active connection, so use the provider to get a new net.Conn, and cache it.
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	newConn, err := d.provider.GetConn()
+	if err != nil {
+		return nil, err
+	}
+	d.conn = newConn
+	return newConn, nil
+}
+
+func (d *TCPWriter) closeConn() error {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.conn != nil {
+		err := d.conn.Close()
+		d.conn = nil
+		return err
+	}
+	return nil
+}
+
+func (d *TCPWriter) closed() bool {
+	select {
+	case <-d.closedChan:
+		return true
+	default:
+		return false
+	}
+}
+
+// Close will close any existing client connections and shuts down the writer from any future writes.
+func (d *TCPWriter) Close() error {
+	close(d.closedChan)
+	return d.closeConn()
+}
+
+func zerologSerializer(metadata LogEnvelopeMetadata) envelopeSerializerFunc {
+	// create a new top level logger with no a scratch output since each
```

Comment:
nit: grammar

```
// create a new top-level logger that writes to ioutil.Discard since each
// serialization will write to its own local buffer instead of a single writer
```

Reply:
🤦‍♂️ fixed